Migrate forloop directional tests to OptimizerInfo

janeyx99 opened the pull request:
This PR is another step towards modernizing our optimizer tests by tackling the simplest foreach tests. The replaced tests are now removed in `test/optim/test_optim.py`.

**Changes in coverage?** Yes! 
- This PR _decreases_ coverage (!!!!) by only checking the direction on the forloop implementations vs both the forloop and foreach. Why? I believe it should be sufficient to check the forloop only, as the foreach parity is already checked in the `foreach_matches_forloop` test.
- This PR also _increases_ coverage for SparseAdam with contiguous params on CUDA, which was previously forbidden due to an old old bug that has since been fixed.

What will it take to fully remove `test_basic_cases`?
- We need to flavor the tests with LRSchedulers
- Testing for param groups --> which all just distinguish between lrs.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117410

janeyx99 committed:
Migrate forloop directional tests to OptimizerInfo

[ghstack-poisoned]

janeyx99 commented on torch/testing/_internal/common_optimizers.py:
```diff
@@ -508,7 +507,7 @@ def optim_inputs_func_lbfgs(device=None):
         OptimizerInput(params=None, kwargs={"lr": 0.01}, desc="non-default lr"),
         OptimizerInput(
-            params=None, kwargs={"tolerance_grad": math.inf}, desc="tolerance_grad"
+            params=None, kwargs={"tolerance_grad": 1e-6}, desc="tolerance_grad"
```
If it's inf, lbfgs will think it should terminate as it's the most optimal already.

janeyx99 committed:
Update on "Migrate forloop directional tests to OptimizerInfo"




[ghstack-poisoned]

janeyx99 committed:
Update on "Migrate forloop directional tests to OptimizerInfo"




[ghstack-poisoned]

janeyx99 commented on test/optim/test_optim.py:
```diff
@@ -1311,12 +1092,12 @@ def test_asgd(self):
             self._test_complex_optimizer(
                 lambda params: ASGD(
-                    [params], maximize=True, weight_decay=0.9, foreach=foreach
+                    [params], maximize=True, weight_decay=0.1, foreach=foreach
```
A quick GitHub grep shows that weight_decay of 0.1 is much much more common than 0.9, since people don't want to overregularize the weights.

janeyx99 committed:
Update on "Migrate forloop directional tests to OptimizerInfo"

This PR is another step towards modernizing our optimizer tests by tackling the simplest foreach tests. The replaced tests are now removed in `test/optim/test_optim.py`.

**Changes in coverage?** Yes! 
- This PR _decreases_ coverage (!!!!) by only checking the direction on the forloop implementations vs both the forloop and foreach. Why? I believe it should be sufficient to check the forloop only, as the foreach parity is already checked in the `foreach_matches_forloop` test.
- This PR also _increases_ coverage for SparseAdam with contiguous params on CUDA, which was previously forbidden due to an old old bug that has since been fixed.

What will it take to fully remove `test_basic_cases`?
- We need to flavor the tests with LRSchedulers
- Testing for param groups




[ghstack-poisoned]

janeyx99 commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:
```diff
@@ -296,12 +296,8 @@ Tensor& add_out_dense_sparse_cuda(Tensor& r_, const Tensor& dense, const SparseT
   Tensor values = sparse._values().to(commonDtype);
 
-  if (is_same_tensor(r, dense_buffer)) {
-    TORCH_CHECK(r_.is_contiguous(), "add: CUDA dense-sparse addition with a non-contiguous output tensor does not work; shout if you need it (see https://github.com/pytorch/pytorch/issues/1521 )");
```
this issue has been fixed!

janeyx99 committed:
Update on "Migrate forloop directional tests to OptimizerInfo"

This PR is another step towards modernizing our optimizer tests by tackling the simplest foreach tests. The replaced tests are now removed in `test/optim/test_optim.py`.

**Changes in coverage?** Yes! 
- This PR _decreases_ coverage (!!!!) by only checking the direction on the forloop implementations vs both the forloop and foreach. Why? I believe it should be sufficient to check the forloop only, as the foreach parity is already checked in the `foreach_matches_forloop` test.
- This PR also _increases_ coverage for SparseAdam with contiguous params on CUDA, which was previously forbidden due to an old old bug that has since been fixed.

What will it take to fully remove `test_basic_cases`?
- We need to flavor the tests with LRSchedulers
- Testing for param groups --> which all just distinguish between lrs.




[ghstack-poisoned]

albanD commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:
```diff
@@ -296,12 +296,8 @@ Tensor& add_out_dense_sparse_cuda(Tensor& r_, const Tensor& dense, const SparseT
   Tensor dense_buffer = dense.to(commonDtype);
   Tensor values = sparse._values().to(commonDtype);
 
-  if (is_same_tensor(r, dense_buffer)) {
```
You need to keep this if statement!
Only resize/copy if r is not dense_buffer.

janeyx99 replied:
Oh...yea

albanD commented on test/test_optim.py:
```diff
@@ -56,6 +56,78 @@ def test_errors(self, device, dtype, optim_info):
                 raise NotImplementedError(f"Unknown error type {error_input.error_on}")
 
 
+    @parametrize("on", ["contiguous_params", "noncontiguous_params"])
```
nit: "on" is not a great variable name. contiguous_param. [True, False] would be easier to read.

janeyx99 committed:
Update on "Migrate forloop directional tests to OptimizerInfo"

This PR is another step towards modernizing our optimizer tests by tackling the simplest foreach tests. The replaced tests are now removed in `test/optim/test_optim.py`.

**Changes in coverage?** Yes! 
- This PR _decreases_ coverage (!!!!) by only checking the direction on the forloop implementations vs both the forloop and foreach. Why? I believe it should be sufficient to check the forloop only, as the foreach parity is already checked in the `foreach_matches_forloop` test.
- This PR also _increases_ coverage for SparseAdam with contiguous params on CUDA, which was previously forbidden due to an old old bug that has since been fixed.

What will it take to fully remove `test_basic_cases`?
- We need to flavor the tests with LRSchedulers
- Testing for param groups --> which all just distinguish between lrs.




[ghstack-poisoned]

janeyx99 committed:
Update on "Migrate forloop directional tests to OptimizerInfo"

This PR is another step towards modernizing our optimizer tests by tackling the simplest foreach tests. The replaced tests are now removed in `test/optim/test_optim.py`.

**Changes in coverage?** Yes! 
- This PR _decreases_ coverage (!!!!) by only checking the direction on the forloop implementations vs both the forloop and foreach. Why? I believe it should be sufficient to check the forloop only, as the foreach parity is already checked in the `foreach_matches_forloop` test.
- This PR also _increases_ coverage for SparseAdam with contiguous params on CUDA, which was previously forbidden due to an old old bug that has since been fixed.

What will it take to fully remove `test_basic_cases`?
- We need to flavor the tests with LRSchedulers
- Testing for param groups --> which all just distinguish between lrs.




[ghstack-poisoned]

albanD reviewed (approved):
SGTM!

janeyx99 commented:
@pytorchbot merge

janeyx99 committed:
Update on "Migrate forloop directional tests to OptimizerInfo"

This PR is another step towards modernizing our optimizer tests by tackling the simplest foreach tests. The replaced tests are now removed in `test/optim/test_optim.py`.

**Changes in coverage?** Yes! 
- This PR _decreases_ coverage (!!!!) by only checking the direction on the forloop implementations vs both the forloop and foreach. Why? I believe it should be sufficient to check the forloop only, as the foreach parity is already checked in the `foreach_matches_forloop` test.
- This PR also _increases_ coverage for SparseAdam with contiguous params on CUDA, which was previously forbidden due to an old old bug that has since been fixed.

What will it take to fully remove `test_basic_cases`?
- We need to flavor the tests with LRSchedulers
- Testing for param groups --> which all just distinguish between lrs.




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

janeyx99 committed:
Update on "Migrate forloop directional tests to OptimizerInfo"

This PR is another step towards modernizing our optimizer tests by tackling the simplest foreach tests. The replaced tests are now removed in `test/optim/test_optim.py`.

**Changes in coverage?** Yes! 
- This PR _decreases_ coverage (!!!!) by only checking the direction on the forloop implementations vs both the forloop and foreach. Why? I believe it should be sufficient to check the forloop only, as the foreach parity is already checked in the `foreach_matches_forloop` test.
- This PR also _increases_ coverage for SparseAdam with contiguous params on CUDA, which was previously forbidden due to an old old bug that has since been fixed.

What will it take to fully remove `test_basic_cases`?
- We need to flavor the tests with LRSchedulers
- Testing for param groups --> which all just distinguish between lrs.




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge
